src/main/java/com/nutomic/syncthingandroid/fragments/SelectFolderFragment.java

```diff
@@ -0,0 +1,108 @@
+
+import android.database.Cursor;
+import android.net.Uri;
+import android.os.AsyncTask;
+import android.provider.MediaStore;
+import android.util.Log;
+import android.view.View;
+import android.widget.AdapterView;
+
+import com.nutomic.syncthingandroid.activities.SyncthingActivity;
+import com.nutomic.syncthingandroid.syncthing.RestApi;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.List;
+
+public class SelectFolderFragment extends FoldersFragment {
+
+    private static final String TAG = "SelectFolderFragment";
+    private ArrayList<Uri> mFiles;
+    private boolean copying = false;
+
+    @Override
+    public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
+        // Currently only handle one copy request per intent. So the UI will remain responsive
+        // but won't really do anything after the first click.
+        if (!copying) {
+            copying = true;
+            RestApi.Folder destFolder = getItemFolder(i);
+            CopyFilesTaskParams copyFilesTaskParams = new CopyFilesTaskParams(mFiles, destFolder);
+            new CopyFilesTask().execute(copyFilesTaskParams);
+        }
+    }
+
+    public void setFiles(ArrayList<Uri> files) {
+        mFiles = files;
+    }
+
+    private static class CopyFilesTaskParams {
+        public final List<Uri> sourceUris;
+        public final RestApi.Folder destFolder;
+        CopyFilesTaskParams(List<Uri> sourceUris, RestApi.Folder destFolder) {
+            this.sourceUris = sourceUris;
+            this.destFolder = destFolder;
+        }
+    }
+
+    private class CopyFilesTask extends AsyncTask<CopyFilesTaskParams, Void, Void> {
+        @Override
+        protected Void doInBackground(CopyFilesTaskParams... params) {
+            List<Uri> sourceUris = params[0].sourceUris;
+            RestApi.Folder destFolder = params[0].destFolder;
+            for (Uri uri : sourceUris) {
+                File sourceFile = new File(getRealPathFromURI(uri));
+                File destFile = new File(destFolder.path, sourceFile.getName());
+                try {
+                    copyFile(sourceFile, destFile);
+                }
+                catch (IOException e) {
+                    Log.e(TAG, "Failed copying file", e);
+                }
+            }
+
+            SyncthingActivity activity = (SyncthingActivity) getActivity();
+            activity.getApi().onFolderFileChange(destFolder.id, destFolder.path);
+            activity.finish();
+            return null;
+        }
+
+        // Copied from http://stackoverflow.com/a/9989900/943814
+        private String getRealPathFromURI(Uri contentURI) {
+            String result;
+            Cursor cursor = getActivity().getContentResolver()
+                    .query(contentURI, null, null, null, null);
```

Comment:
I just dumped this cursor for an image from the gallery, and got all kinds of data (`DatabaseUtils.dumpCursorToString(cursor)`)
```:
08-26 20:20:01.061    5899-6488/com.nutomic.syncthingandroid.debug D/SelectFolderFragment﹕ >>>>> Dumping cursor android.content.ContentResolver$CursorWrapperInner@3ab483f7
    0 {
    _id=20502
    _data=/storage/emulated/0/DCIM/Camera/IMG_20150824_164618.jpg
    _size=2480077
    _display_name=IMG_20150824_164618.jpg
    mime_type=image/jpeg
    title=IMG_20150824_164618
    date_added=xxx
    date_modified=xxx
    description=null
    picasa_id=null
    isprivate=null
    latitude=xxx
    longitude=xxx
    datetaken=xxx
    orientation=0
    mini_thumb_magic=1245638041573583436
    bucket_id=-1739773001
    bucket_display_name=Camera
    width=3200
    height=2368
    }
    <<<<<
```
So I don't think the hack is necessary. And you should probably pass a selection (instead of null).